### PR TITLE
use a string instead of macro for version number

### PR DESCRIPTION
**Pull request by kroening**
The key benefit is the reduction in build time, as now many more files are amenable to cached builds.

### CODE REVIEW

**Review by smowton — COMMENTED** [review]
I think for cmake you don't want `add_dependencies` any more, but rather to add `${CMAKE_CURRENT_BINARY_DIR}/version.cpp` to the `add_library` command

File: src/util/CMakeLists.txt [hunk 1]
```diff
@@ -28,18 +28,18 @@ else()
...
-    configure_file(\${CUR}/version.h.in version.h)
+    configure_file(\${CUR}/version.cpp.in version.cpp)
     "
   )
 endif()
 
-file(WRITE ${CMAKE_CURRENT_BINARY_DIR}/version.h.in
-  "\#define CBMC_VERSION \"@CBMC_RELEASE@ (@GIT_INFO@)\"\n")
+file(WRITE ${CMAKE_CURRENT_BINARY_DIR}/version.cpp.in
+  "extern const char CBMC_VERSION[]=\"@CBMC_RELEASE@ (@GIT_INFO@)\";\n")
```
Comment (smowton) [inline src/util/CMakeLists.txt]: Remove `extern` from the definition, and use `const char *` to match the declaration?

Comment (kroening) [reply src/util/CMakeLists.txt]: The extern is needed, or clang will complain that the symbol is unused.

**Review by peterschrammel — APPROVED** [review]
Fixed.

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 8ef30b46772703faa1b726c84cef72ea2dc99d94).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/80517653

File: jbmc/src/janalyzer/janalyzer_parse_options.cpp [hunk 2]
```diff
@@ -60,7 +60,7 @@ Author: Daniel Kroening, kroening@kroening.com
 janalyzer_parse_optionst::janalyzer_parse_optionst(int argc, const char **argv)
   : parse_options_baset(JANALYZER_OPTIONS, argc, argv),
     messaget(ui_message_handler),
-    ui_message_handler(cmdline, "JANALYZER " CBMC_VERSION)
+    ui_message_handler(cmdline, std::string("JANALYZER ")+CBMC_VERSION)
```
Comment (tautschnig) [inline jbmc/src/janalyzer/janalyzer_parse_options.cpp]: Nit picking: clang-format will ask for space around `+` (applies across this PR)

**Review by tautschnig — APPROVED** [review]
Nice!

File: jbmc/src/janalyzer/janalyzer_parse_options.cpp (hunk 2 above)
Comment (kroening) [reply jbmc/src/janalyzer/janalyzer_parse_options.cpp]: done

File: src/util/version.h [hunk 3]
```diff
@@ -0,0 +1 @@
+extern const char *CBMC_VERSION;
```
Comment (peterschrammel) [inline src/util/version.h:1]: cpplint complains about the missing copyright header and header guard

**Review by chrisr-diffblue — APPROVED** [review]

**Pull request merged**